From 4533c285c2aedce6d4434d7b877066de3b1ecb33 Mon Sep 17 00:00:00 2001
From: Georgi Djakov <georgi.djakov@linaro.org>
Date: Thu, 25 Aug 2016 18:43:35 +0300
Subject: [PATCH 52/69] PM / OPP: Update the voltage tolerance when adjusting
 the OPP

When the voltage is adjusted, the voltage tolerance is not updated.
This can lead to situations where the voltage min value is greater
than the voltage max value. The final result is triggering a BUG()
in the regulator core.
Fix this by updating the voltage tolerance values too.

Signed-off-by: Georgi Djakov <georgi.djakov@linaro.org>
---
 drivers/opp/core.c | 5 +++++
 1 file changed, 5 insertions(+)

--- a/drivers/opp/core.c
+++ b/drivers/opp/core.c
@@ -2143,6 +2143,7 @@ int dev_pm_opp_adjust_voltage(struct dev
 	struct opp_table *opp_table;
 	struct dev_pm_opp *tmp_opp, *opp = ERR_PTR(-ENODEV);
 	int r = 0;
+	unsigned long tol;
 
 	/* Find the opp_table */
 	opp_table = _find_opp_table(dev);
@@ -2172,8 +2173,17 @@ int dev_pm_opp_adjust_voltage(struct dev
 		goto adjust_unlock;
 
 	opp->supplies->u_volt = u_volt;
-	opp->supplies->u_volt_min = u_volt_min;
-	opp->supplies->u_volt_max = u_volt_max;
+
+	tol = u_volt * opp_table->voltage_tolerance_v1 / 100;
+	if ( u_volt_min == u_volt )
+		opp->supplies->u_volt_min = u_volt - tol;
+	else
+		opp->supplies->u_volt_min = u_volt_min;
+
+	if ( u_volt_max == u_volt )
+		opp->supplies->u_volt_max = u_volt + tol;
+	else
+		opp->supplies->u_volt_max = u_volt_max;
 
 	dev_pm_opp_get(opp);
 	mutex_unlock(&opp_table->lock);
